# Regressão Linear Polinomial
#### Aplicando o algoritmo de regressão linear polinomial em uma base de dados não linear de preço de casas.

In [1]:
# importando a biblioteca pandas do python
import pandas as pd

In [2]:
# importando a base de dados no dataframe
dataframe = pd.read_csv('house_prices.csv', encoding = 'utf-8', sep = ',')

In [3]:
# separando as variáveis x e y
x = dataframe.iloc[:, 3:19].values
y = dataframe.iloc[:, 2].values

# lembrando que temos mais de um variável x, mas não são todas presentes na base de dados

In [11]:
# visualizando as dimensões das variáveis x e y
print('Dimensão de x: {}'.format(x.shape))
print('Dimensão de y: {}'.format(y.shape))

Dimensão de x: (21613, 16)
Dimensão de y: (21613,)


In [5]:
# importando a biblioteca sklearn do python
from sklearn.model_selection import train_test_split

In [6]:
# separando o a base de dados para treinamento e para teste
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 0)

In [7]:
# visualizando a dimensão da base de dados 'x_train'
print(x_train.shape)

(15129, 16)


In [8]:
# visualizando a dimensão da base de dados 'x_test'
print(x_test.shape)

(6484, 16)


In [9]:
# visualizando a dimensão da base de dados 'y_train'
print(y_train.shape)

(15129,)


In [10]:
# visualizando a dimensão da base de dados 'y_test'
print(y_test.shape)

(6484,)


In [13]:
# importando a biblioteca sklearn do python
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

In [14]:
# criando o objeto 'poly' responsável por elevar as variáveis x a 4º potência
poly = PolynomialFeatures(degree = 4)

In [15]:
# elevando as variáveis x a 4º potência
x_train_poly = poly.fit_transform(x_train)
x_test_poly = poly.fit_transform(x_test)

In [16]:
# criando o classificador 'regressor'
regressor = LinearRegression()

In [17]:
# treinando o classificador
regressor.fit(x_train_poly, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [19]:
# visualizando o quanto o modelo criado se adaptou aos dados
score = regressor.score(x_train_poly, y_train)
# visualizando a porcentagem
print(score)

0.8550388465809267


### Realizando Predições

In [26]:
# aplicando a base de dados de teste para obter resultados
previsoes = regressor.predict(x_test_poly)

In [27]:
# visualizando as previsões feitas
print(previsoes)

[ 403127.22894001 2024749.14025784  546325.01213074 ...  534665.47343349
  244341.65425205   99821.1396122 ]


### Erro Médio

In [28]:
# importando a biblioteca sklearn do python
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_test, previsoes)

In [29]:
# visualizando o erro médio
print(mae)

150554.90163297008


A variação de erro para essa de base é de mais ou menos 150554.90163297008, um valor muito alto.

### Alguma Dúvida? Entre em Contato Comigo:

- [Me envie um e-mail](mailto:alyssonmachado388@gmail.com);